veracruz-client/src/veracruz_client.rs

```diff
@@ -11,26 +11,96 @@
 
 use crate::error::VeracruzClientError;
 use log::{error, info};
+use mbedtls::alloc::List;
 use policy_utils::{parsers::enforce_leading_backslash, policy::Policy, Platform};
-use rustls::{Certificate, ClientConnection, PrivateKey};
-use rustls_pemfile;
 use std::{
     convert::TryFrom,
     io::{Read, Write},
     path::Path,
-    str::from_utf8,
+    sync::{Arc, Mutex},
 };
 use veracruz_utils::VERACRUZ_RUNTIME_HASH_EXTENSION_ID;
 use webpki;
 
-#[derive(Debug)]
+/// VeracruzClient struct. The remote_session_id is shared between
+/// VeracruzClient and InsecureConnection so that it is available from
+/// VeracruzClient methods and can also be updated by the
+/// InsecureConnection methods invoked by mbedtls.
 pub struct VeracruzClient {
-    tls_connection: ClientConnection,
-    remote_session_id: Option<u32>,
+    tls_connection: mbedtls::ssl::Context<InsecureConnection>,
+    remote_session_id: Arc<Mutex<Option<u32>>>,
```

Comment:
Why is the `Arc<Mutex` now required for this?

Reply:
The two functions that update the `remote_session_id` aren't really going to be in separate threads, I think, but because one is called via C there is no way for the Rust compiler to do any kind of memory check, so I have to choose between `unsafe` and `Arc<Mutex>`. Since it's not performance-critical, I went for the latter.